libmamba/include/mamba/util/flat_bool_expr_tree.hpp

```diff
@@ -0,0 +1,738 @@
+        return push_operator_impl(op);
+    }
+
+    template <typename V, typename O>
+    void PostfixParser<V, O>::push_operator(operator_type&& op)
+    {
+        return push_operator_impl(std::move(op));
+    }
+
+    template <typename V, typename O>
+    void PostfixParser<V, O>::finalize()
+    {
+        if (m_orphans.size() != 1)
+        {
+            throw std::invalid_argument("Incomplete expression");
+        }
+    }
+
+    template <typename V, typename O>
+    auto PostfixParser<V, O>::tree() const& -> const tree_type&
+    {
+        return m_tree;
+    }
+
+    template <typename V, typename O>
+    auto PostfixParser<V, O>::tree() && -> tree_type&&
+    {
+        return std::move(m_tree);
+    }
+
+    /***********************************
+     *  Implementation of InfixParser  *
+     ***********************************/
+
+    template <typename V, typename O, typename C>
+    InfixParser<V, O, C>::InfixParser(const operator_precedence_type& cmp)
+        : m_op_cmp(cmp)
+    {
+    }
+
+    template <typename V, typename O, typename C>
+    InfixParser<V, O, C>::InfixParser(operator_precedence_type&& cmp)
+        : m_op_cmp(std::move(cmp))
+    {
+    }
+
+    template <typename V, typename O, typename C>
+    template <typename T>
+    void InfixParser<V, O, C>::stack_push(T&& elem)
+    {
+        m_op_stack.push_back(std::forward<T>(elem));
+    }
+
+    template <typename V, typename O, typename C>
+    auto InfixParser<V, O, C>::stack_pop() -> operator_or_parenthesis_type
+    {
+        assert(!stack_empty());
+        auto top = stack_top();
+        m_op_stack.pop_back();
+        return top;
+    }
+
+    template <typename V, typename O, typename C>
+    auto InfixParser<V, O, C>::stack_empty() const -> bool
+    {
+        return m_op_stack.empty();
+    }
+
+    template <typename V, typename O, typename C>
+    auto InfixParser<V, O, C>::stack_top() const -> const operator_or_parenthesis_type&
+    {
+        assert(!stack_empty());
+        return m_op_stack.back();
+    }
+
+    template <typename V, typename O, typename C>
+    auto InfixParser<V, O, C>::stack_top_is_parenthesis() const -> bool
+    {
+        return (!stack_empty()) && std::holds_alternative<LeftParenthesis>(stack_top());
+    }
+
+    template <typename V, typename O, typename C>
+    auto InfixParser<V, O, C>::stack_top_is_op_with_greater_precedence_than(const operator_type& op
+    ) const -> bool
+    {
+        if (stack_empty())
+        {
+            return false;
+        }
+        if (const auto* const op_ptr = std::get_if<operator_type>(&stack_top()))
+        {
+            return m_op_cmp(op, *op_ptr);
+        }
+        return false;
+    }
+
+    template <typename V, typename O, typename C>
+    template <typename Var>
+    void InfixParser<V, O, C>::push_variable_impl(Var&& var)
+    {
+        // Input check
+        if (m_expects_op)
+        {
+            throw std::invalid_argument("Unexpected variable");
```

Comment:
Consider adding the variable in the error report.

Reply:
Same